File: C++/BOJ/11724.cpp
#include<iostream>
#include<vector>
#include<stack>
#include<deque>
using namespace std;

void dfs(vector<bool> *visited, int v, vector<vector<int>> graph)
{
    stack<int>  stack;
    stack.push(v);
    (*visited)[v] = true;
    while (stack.size())
    {
        int i = stack.top();
        stack.pop();
        for (auto j : graph[i])
        {
            if ((*visited)[j] == false)
            {
                stack.push(j);
                (*visited)[j] = true;
            }
        }
    }
}

void bfs(vector<bool> *visited, int v, vector<vector<int>> graph)
{
    deque<int>  queue;
    queue.push_back(v);
    while (queue.size())
    {
        int i = queue.front();
        queue.pop_front();
        for (auto j : graph[i])
        {
            if ((*visited)[j] == false)
            {
                queue.push_back(j);
                (*visited)[j] = true;
            }
        }
    }
}

int main()
{
    int dots, lines;
    cin >> dots>> lines;
    int u, v;
    int count = 0;
    vector<bool> visited;
    vector<vector<int>>  graph;
    for(int i = 0; i <= dots; i++)
    {
        visited.push_back(false);
        graph.push_back({});
    }
    for(int i = 0; i < lines; i++)
    {
        cin>> u >> v;
        graph[u].push_back(v);
        graph[v].push_back(u);
    }
    for(int v = 1; v <= dots; v++)
    {
        if (visited[v] == true) continue;
        //dfs(&visited, v, graph);
        bfs(&visited,v,graph);
        count++;
    }
    cout<<count;
}

File: Python/BOJ/11724.py
import sys
from collections import deque
sys.setrecursionlimit(10000)

def dfs(visited, v, graph):
    stack = [v]
    visited[v] = True
    while stack:
        i = stack.pop()
        for j in graph[i]:
            if not visited[j]:
                stack.append(j)
                visited[j] = True


def bfs(visited, v, graph):
    queue = deque([v])
    visited[v] = True
    while queue:
        i = queue.popleft()
        for j in graph[i]:
            if not visited[j]:
                queue.append(j)
                visited[j] = True


def solution():
    dots, lines = map(int, sys.stdin.readline().split())
    graph = [[] for i in range(dots + 1)]
    visited = [False] * (dots + 1)
    count = 0
    for i in range(lines):
        u, v = map(int, sys.stdin.readline().split())
        graph[u].append(v)
        graph[v].append(u)
    #print(graph)
    for v in range(1, dots + 1):
        if visited[v]: continue
        dfs(visited, v, graph)
        count += 1
    # for v in range(1, dots + 1):
    #     if visited[v] : continue
    #     bfs(visited, v, graph)
    #     count += 1
    print(count)

solution()